src/riak_kv_ts_util.erl

```diff
@@ -437,69 +432,76 @@ explain_query(QueryString) ->
 %%
 %% Have a flexible API because it is a debugging function.
 explain_query(DDL, ?SQL_SELECT{} = Select) ->
-    {ok, SubQueries} = riak_kv_qry_compiler:compile(DDL, Select, 10000),
-    [explain_sub_query(SQ) || SQ <- SubQueries];
+    case riak_kv_qry_compiler:compile(DDL, Select, 10000) of
+        {ok, SubQueries} ->
+            {_Total, Rows} = lists:foldl(fun index_subquery/2, {1,[]}, SubQueries),
+            lists:reverse(Rows);
+        Error ->
+            Error
+    end;
 explain_query(DDL, QueryString) ->
     {ok, Select} = explain_compile_query(QueryString),
     explain_query(DDL, Select).
 
 %%
+index_subquery(SQ, {Idx, Acc}) ->
+    Row = explain_sub_query(Idx, SQ),
+    {Idx + 1, [Row | Acc]}.
+
+%%
 explain_compile_query(QueryString) ->
     {ok, Q} = riak_ql_parser:parse(riak_ql_lexer:get_tokens(QueryString)),
     build_sql_record(select, Q, undefined).
 
 %%
-explain_sub_query(#riak_select_v1{ 'WHERE' = SubQueryWhere }) ->
+explain_sub_query(Index, ?SQL_SELECT{ 'WHERE' = SubQueryWhere }) ->
     {_, StartKey1} = lists:keyfind(startkey, 1, SubQueryWhere),
     {_, EndKey1} = lists:keyfind(endkey, 1, SubQueryWhere),
     {_, Filter} = lists:keyfind(filter, 1, SubQueryWhere),
-    explain_query_keys(StartKey1, EndKey1, Filter).
+    StartKey2 = string:join([key_to_string(Key) || Key <- StartKey1], ", "),
+    EndKey2 = string:join([key_to_string(Key) || Key <- EndKey1], ", "),
+    StartInclusive = lists:keymember(start_inclusive, 1, StartKey1),
+    EndInclusive = lists:keymember(end_inclusive, 1, EndKey1),
+    [Index, StartKey2, StartInclusive, EndKey2, EndInclusive, filter_to_string(Filter)].
 
 %%
-explain_query_keys(StartKey1, EndKey1, Filter) ->
-    StartKey2 = [[key_element_to_string(V), $/] || {_,_,V} <- StartKey1],
-    EndKey2 = [[key_element_to_string(V), $/] || {_,_,V} <- EndKey1],
-    case lists:keyfind(start_inclusive, 1, StartKey1) of
-        {start_inclusive,true} ->
-            StartKey3 = [">= ", StartKey2];
-        _ ->
-            StartKey3 = [">  ", StartKey2]
-    end,
-    case lists:keyfind(end_inclusive, 1, EndKey1) of
-        {end_inclusive,true} ->
-            EndKey3 = ["<= ", EndKey2];
-        _ ->
-            EndKey3 = ["<  ", EndKey2]
-    end,
-    {StartKey3, EndKey3, Filter}.
+key_element_to_string(V) when is_binary(V) -> binary_to_list(V);
```

Comment:
I'm wondering if all this string manipulation stuff should go in the riak_ql to string module that converts DDLs to strings.